pkg/federate/kubefed/resource_test.go

```diff
@@ -0,0 +1,389 @@
+			Expect(found).To(BeTrue(), "Hierarchy of fields missing!")
+			Expect(mapObj["spec"]).To(BeEquivalentTo(targetResource.Object["spec"]))
+			Expect(mapObj["metadata"]).To(BeEquivalentTo(targetResource.Object["metadata"]))
+		})
+
+		It("the returned resource's placement should be an empty matcher", func() {
+			// This test implicitly tests that the nested hierarchy of fields exists
+			expected, _ = createFederatedResource(scheme, resource)
+
+			mapObj, found, err := getMatchLabelsField(expected)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(found).To(BeTrue(), "Hierarchy of fields missing!")
+			Expect(mapObj).To(BeEmpty())
+		})
+
+		It("the returned resource's namespace should be the input's namespace", func() {
+			expected, _ = createFederatedResource(scheme, resource)
+			Expect(expected.GetNamespace()).
+				To(Equal(resource.GetNamespace()))
+		})
+	})
+}
+
+func testDistribute() {
+	var (
+		resource      *corev1.Pod
+		fedResource   *unstructured.Unstructured
+		f             *federator
+		initObjs      []runtime.Object
+		clusterNames  []string
+		scheme        *runtime.Scheme
+		clientPatcher client.Client
+	)
+
+	BeforeEach(func() {
+		resource = newPod("test-pod", "nginx")
+		scheme = runtime.NewScheme()
+	})
+
+	AfterEach(func() {
+		resource = nil
+		fedResource = nil
+		f = nil
+		initObjs = []runtime.Object{}
+		clusterNames = []string{}
+		scheme = nil
+		clientPatcher = nil
+	})
+
+	JustBeforeEach(func() {
+		f = newFederatorWithScheme(scheme, clientPatcher, initObjs...)
+	})
+
+	When("the resource type scheme has NOT been added to the type registry", func() {
+		It("should throw an error", func() {
+			err := f.Distribute(resource, clusterNames...)
+			Expect(err).To(HaveOccurred())
+		})
+	})
+
+	When("the resource type scheme has been added to the type registry", func() {
+		BeforeEach(func() {
+			err := corev1.AddToScheme(scheme)
+			Expect(err).ToNot(HaveOccurred())
+			fedResource, _ = createFederatedResource(scheme, resource)
+		})
+
+		It("should return no error on success", func() {
+			err := f.Distribute(resource, clusterNames...)
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		When("the resource is already in the Kube API", func() {
+			expectedImage := "apache"
+
+			BeforeEach(func() {
+				initObjs = append(initObjs, fedResource)
+				resource = newPod("test-pod", expectedImage)
+			})
+
+			It("should update the resource", func() {
+				_ = f.Distribute(resource, clusterNames...)
+				fedPod, err := getResourceFromAPI(f.kubeFedClient, fedResource)
+				Expect(err).ToNot(HaveOccurred())
+				fedPodObj, _, err := getTemplateField(fedPod)
+				Expect(err).ToNot(HaveOccurred())
+				expectObj, _, err := getTemplateField(fedResource)
+				Expect(err).ToNot(HaveOccurred())
+
+				Expect(fedPodObj["spec"]).ToNot(BeEquivalentTo(expectObj["spec"]))
+
+				mapObj, _, _ := unstructured.NestedSlice(
+					fedPod.Object,
+					ctlutil.SpecField, ctlutil.TemplateField, "spec", "containers",
+				)
```

Comment:
For clarity, to follow actual vs expected semantics, some suggested var name changes:

fedPod -> actualResource
fedPodObj -> actualTemplate
mapObj -> actualContainers
image -> actualImage
fedResource -> expResource
expectObj -> expTemplate